sub/filter.c

```diff
@@ -0,0 +1,240 @@
+ * with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdlib.h>
+#include <assert.h>
+#include <string.h>
+#include "common/common.h"
+#include "common/msg.h"
+#include "sd.h"
+
+// Filter for removing subtitle additions for deaf or hard-of-hearing (SDH)
+// This is for English, but may in part work for others too.
+// The intention is that it can be on always so may not remove
+// all SDH parts.
+// This is for filtering ASS encoded subtitles
+//
+// The incoming ass line is overwritten in place (starting at text position)
+// If result is empty line the first postion of the ass line will be set to NULL.
+//
+// Future optimization could be to include state so format and n_ignored
+// could be done once.
+//
+// Future enhancement could be to add "- " when a speaker label
+// is removed. Need to check what is used in current file as some
+// use "-" and some nothing at all.
+void filter_SDH(struct sd *sd, char * format, int n_ignored, char * ass, int filter_harder)
+{
+    char * p = ass;
+
+    if (!format)
+        return;
+
+    int comma = 0;
+    // scan format line to find where the text is
+    for (char *c = format; *c; c++) {
+        if (*c == ',') {
+            comma++;
+            if (strncasecmp(c + 1, "Text", 4) == 0)
+                break;
+        }
+    }
+    // if preprocessed line some commas are skipped
+    comma -= n_ignored;
+
+    // locate text
+    for (int k = 0; k < comma; k++) {
+        while (*p && *p != ',')
+            p++;
+        if (*p == ',')
+            p++;
+    }
+    // p is now at start of text
+    char * ass_text_start = p;
+    char * t = p;   // current writing position
+
+    // go through the lines in the text
+    // they are separated by \N
+    for (;*p;) {
+        // now at beginning of a line
+        char * text_with_ass_formatting_start = t;
+        // skip leading ass formatting like bold or italic
+        while (*p == '{') {
+            *t++ = *p++;
+            for (; *p && *p != '}'; *t++ = *p++);
+            if (*p)
+                *t++ = *p++;
+        }
+        char * line_start = p;
+        char * text_start = t;
+        // move past leading - 
+        if (*p == '-') {
+            *t++ = *p++;
+        }
+        while (*p == ' ')
+            *t++ = *p++;
+        char * text_after_hyphon_start = t;
+        // check for speaker label, like MAN:
+        // normal subtitles may include mixed case text with : after so
+        // only upper case is accepted and lower case l which for some looks like upper case I
+        char * c;
+        if (filter_harder) {
+            // include lower case also
+            for (c = p; *c && *c != ':' && ((*c >= 'A' && *c <= 'Z') || (*c >= 'a' && *c <= 'z') ||
+                                            (*c >= '0' && *c <= '9') ||
+                                            *c == ' ' || *c == '\'' ||
+                                            *c == '#' || *c == '.' || *c == ',') ; c++);
```

Comment:
I have now done some timing tests (which I should have done before) and ctype functions are as fast so it will not slow it down. I will change to ctype functions.